scvi/model/base/_pyromixin.py

```diff
@@ -0,0 +1,415 @@
+            if any(
+                f.name == self.module.model.list_obs_plate_vars()["name"]
+                for f in site["cond_indep_stack"]
+            )
+        }
+        return obs_plate
+
+    def _posterior_samples_minibatch(
+        self, use_gpu: bool = True, batch_size: int = 128, **sample_kwargs
+    ):
+        """
+        Generate samples of the posterior distribution of each parameter, separating local (minibatch) variables
+        and global variables, which is necessary when performing minibatch inference.
+
+        Note for developers: requires model class method which lists observation/minibatch plate
+        variables (self.module.model.list_obs_plate_vars()).
+
+        Parameters
+        ----------
+        use_gpu
+            Bool, use gpu?
+
+        Returns
+        -------
+        dictionary {variable_name: [array with samples in 0 dimension]}
+
+        """
+
+        gpus, device = parse_use_gpu_arg(use_gpu)
+
+        if batch_size is None:
+            batch_size = self.adata.n_obs
+        train_dl = AnnDataLoader(self.adata, shuffle=False, batch_size=batch_size)
+        # sample local parameters
+        i = 0
+        for tensor_dict in track(
+            train_dl,
+            style="tqdm",
+            description="Sampling local variables, batch: ",
+        ):
+            args, kwargs = self.module._get_fn_args_from_batch(tensor_dict)
+            args = [a.to(device) for a in args]
+            kwargs = {k: v.to(device) for k, v in kwargs.items()}
+            self.to_device(device)
+
+            if i == 0:
+                sample_kwargs_obs_plate = sample_kwargs.copy()
+                sample_kwargs_obs_plate[
+                    "return_sites"
+                ] = self._check_obs_plate_return_sites(sample_kwargs)
+                sample_kwargs_obs_plate["show_progress"] = False
+                obs_plate = self._find_plate_dimension(args, kwargs)
+                obs_plate_dim = list(obs_plate.values())[0]
+                samples = self._get_posterior_samples(
+                    args, kwargs, **sample_kwargs_obs_plate
+                )
+            else:
+                samples_ = self._get_posterior_samples(
+                    args, kwargs, **sample_kwargs_obs_plate
+                )
+
+                samples = {
+                    k: np.array(
+                        [
+                            np.concatenate(
+                                [samples[k][j], samples_[k][j]],
+                                axis=obs_plate_dim,
+                            )
+                            for j in range(
+                                len(samples[k])
+                            )  # for each sample (in 0 dimension
+                        ]
+                    )
+                    for k in samples.keys()  # for each variable
+                }
+            i += 1
+
+        # sample global parameters
+        tensor_dict = next(iter(train_dl))
```

Comment:
for full batch, this will copy yet again to GPU, can't you `tensor_dict` without redfining it because it's defined in the previous loop?

Reply:
(and subsequently, the already defined args and kwargs in scope)